### PR TITLE
gh-89770: Implement PEP-678 - Exception notes

### PR DESCRIPTION
**Pull request by iritkatriel**

### CODE REVIEW

File: Doc/library/exceptions.rst [hunk 1]
```diff
@@ -126,13 +126,20 @@ The following exceptions are used mostly as base classes for other exceptions.
...
+      Add the string ``note`` to the exception's notes which appear in the standard
+      traceback after the exception string. A :exc:`TypeError` is raise if ``note``
+      is not a string.
 
-   .. versionadded:: 3.11
+      .. versionadded:: 3.11
+
+   .. attribute:: __notes__
+
+      A tuple of the notes of this exception, which were added with :meth:`add_note`.
```
Comment (iritkatriel) [inline Doc/library/exceptions.rst]: Need to say that it may not exist.

File: Objects/exceptions.c [hunk 2]
```diff
@@ -194,25 +194,34 @@ _PyBaseExceptionObject_cast(PyObject *exc)
...
+        Py_DECREF(new_notes);
+    }
+    PyObject *notes = PyObject_GetAttr(self, &_Py_ID(__notes__));
+    if (notes == NULL) {
+        return NULL;
     }
-    assert(PyUnicode_CheckExact(note)); 
-    if (PyList_Append(self->notes, note) < 0) {
+    if (PyList_Append(notes, note) < 0) {
+        Py_DECREF(notes);
```
Comment (gvanrossum) [inline Objects/exceptions.c:227]: This probably needs to be slightly nicer -- if `__notes__` isn't a list this raises `SystemError("bad argument to internal function")`. (Unless there's no way to assign to it? But there's a way to delete it. Somehow I can't find the descriptor/property that gives access to it from user code.)

File: Doc/library/exceptions.rst (hunk 1 above)
Comment (gvanrossum) [reply Doc/library/exceptions.rst]: And that it's now a list.

File: Lib/test/test_exceptions.py [hunk 3]
```diff
@@ -538,26 +538,28 @@ def testAttributes(self):
...
+                del e.__notes__
+                self.assertFalse(hasattr(e, '__notes__'))
+
+                e.add_note("Our Note")
+                self.assertEqual(e.__notes__, ["Our Note"])
 
-                e.__note__ = None
-                self.assertIsNone(e.__note__)
+                e.__notes__ = 42
+                self.assertEqual(e.__notes__, 42)
```
Comment (gvanrossum) [inline Lib/test/test_exceptions.py:571]: I'd like to see a test for what `e.add_note(...)` does in this state.

**Review by gvanrossum — COMMENTED** [review]
Where's the descriptor for `"__notes__"`?

File: Objects/exceptions.c (hunk 2 above)
Comment (iritkatriel) [reply Objects/exceptions.c:227]: You're right, I will add a check that it's a list. 

This version implement's [the SC's suggestion](https://discuss.python.org/t/pep-678-enriching-exceptions-with-notes/13374/69) where there is no descriptor, so we only allocate space for ``__notes__`` if you use them.   For the typical user it means to call add_note() and see the notes in the traceback.   If the ``__note__`` is not a sequence it is ignored, and any item in it which is not a string is skipped. So if you mess with it directly it shouldn't blow up.

File: Lib/test/test_exceptions.py [hunk 4]
```diff
@@ -561,6 +561,9 @@ def test_notes(self):
                 e.__notes__ = 42
                 self.assertEqual(e.__notes__, 42)
 
+                e.add_note("will not work")
```
Comment (gvanrossum) [inline Lib/test/test_exceptions.py]: Hm, I'd expect it to fail in this case, not silently swallow the note.

**Review by gvanrossum — COMMENTED** [review]
"Errors should never pass silently."

**Comment by gvanrossum** [comment]
Yeah, I like this!

**Comment by iritkatriel** [comment]
How about error handling when printing the traceback?   Should we print something like "cannot print __notes__: not a sequence", and similarly if a note is not a str?

**Comment by gvanrossum** [comment]
> How about error handling when printing the traceback? Should we print something like "cannot print **notes**: not a sequence", and similarly if a note is not a str?

How about printing `repr(self.__notes__)` if it's not a list, and `str(note)` if an individual note isn't a string? The former makes it clear what it *is*, while the latter provides a feature that Barry has been asking for...

If either of those fails, either raise an exception from `print_exception()` or swallow the error and print a simple message. It probably doesn't really matter, `repr()` or `str()` failing should be extremely obscure.

**Comment by iritkatriel** [comment]
> > How about error handling when printing the traceback? Should we print something like "cannot print **notes**: not a sequence", and similarly if a note is not a str?
> 
> How about printing `repr(self.__notes__)` if it's not a list, and `str(note)` if an individual note isn't a string? The former makes it clear what it _is_, while the latter provides a feature that Barry has been asking for...
> 
> If either of those fails, either raise an exception from `print_exception()` or swallow the error and print a simple message. It probably doesn't really matter, `repr()` or `str()` failing should be extremely obscure.

I'll do something like [what we do for str(exc)](https://github.com/python/cpython/blob/30b5d41fabad04f9f34d603f1ce2249452c18c71/Lib/traceback.py#L173).

**Comment by Zac-HD** [comment]
Heya - I think we should have an explicit test that if you `.split()` an `ExceptionGroup` which already has `__notes__`, it's shallow-copied rather than simply assigned to the new instances.  It'd be pretty confusing to call `.add_note()` on one and have it added to other instances too!

**Comment by iritkatriel** [comment]
> Heya - I think we should have an explicit test that if you `.split()` an `ExceptionGroup` which already has `__notes__`, it's shallow-copied rather than simply assigned to the new instances. It'd be pretty confusing to call `.add_note()` on one and have it added to other instances too!

That's a good point. I added the test and fixed the code to copy the notes in split(), so long as they are a sequence (producing a list).  If people assign arbitrary objects to notes I don't know how to copy them.
Is there a "copy protocol" we can demand?

**Comment by Zac-HD** [comment]
> If people assign arbitrary objects to notes I don't know how to copy them. Is there a "copy protocol" we can demand?

I'd just call `copy.deepcopy()`, and if that doesn't work - well, we only specified support for strings!

**Comment by JelleZijlstra** [comment]
Why not just put `list(other.__notes__)`? I don't see why you'd need to deepcopy.

**Comment by Zac-HD** [comment]
> Why not just put `list(other.__notes__)`? I don't see why you'd need to deepcopy.

You'd get very weird behaviour around `ExceptionGroup.split()` from mutable elements which are now in two notes lists.  This is basically fine since the spec only says to support strings, but if it's easy to avoid having the crazier tricks break only on splitting exceptiongroups I'd like to do so.

**Comment by gvanrossum** [comment]
> > Why not just put `list(other.__notes__)`? I don't see why you'd need to deepcopy.
> 
> You'd get very weird behaviour around `ExceptionGroup.split()` from mutable elements which are now in two notes lists. This is basically fine since the spec only says to support strings, but if it's easy to avoid having the crazier tricks break only on splitting exceptiongroups I'd like to do so.

Hm. In my experience code that deep-copies something is almost always subtly broken. I strongly recommend using a shallow copy here -- `list(exc.__notes__)`, nothing more complicated. People who play games by using mutable notes and actually mutating them will get what they deserve.

**Comment by iritkatriel** [comment]
> Hm. In my experience code that deep-copies something is almost always subtly broken. I strongly recommend using a shallow copy here -- `list(exc.__notes__)`, nothing more complicated. People who play games by using mutable notes and actually mutating them will get what they deserve.


And if list(notes) fails we just copy by reference?

This is fine, but we can't say in the PEP that we support arbitrary stringifiable objects as ``__notes__``.

**Comment by gvanrossum** [comment]
> > Hm. In my experience code that deep-copies something is almost always subtly broken. I strongly recommend using a shallow copy here -- `list(exc.__notes__)`, nothing more complicated. People who play games by using mutable notes and actually mutating them will get what they deserve.
> 
> And if list(notes) fails we just copy by reference?

In that case I would just not create `__notes__` on the new group(s).

> This is fine, but we can't say in the PEP that we support arbitrary stringifiable objects as `__notes__`.

The `str()` of a list looks like its `repr()`, so not what we want for the case of several notes.

IIUC `__notes__` should always be a list (or tuple -- I forget if we landed on one or the other, it seems we picked list?). It's the items in that list that are just arbitrary stringifiables (and shallow copying the list makes that easier since there's one less error to test for).

**Comment by Zac-HD** [comment]
As of https://github.com/python/peps/pull/2456, the PEP explicitly does not specify behaviour if `__notes__` has been modified and is not a list of strings.  However CPython will of course have *some* behaviour if this happens, and I'd suggest the following:

- Attempt list operations on `__notes__` without any error handling.
  - If it's been assigned a tuple and then `.add_note(...)` is called, that should raise the natural error.  
  - If `__notes__` is not an iterable at traceback-display time, let's just skip displaying notes entirely. I'm assuming here that all we do is iterate - i.e. adding no validation code, nor handling for unexpected cases.
- Elements are assumed to be strings
  - shallow-copy with `list(err.__notes__)` when splitting exceptiongroups as Guido suggests
  - use `_safe_string` to stringify non-`str` values when displaying tracebacks, motivated more by error-suppression than supporting arbitrary stringifiable values.

File: Doc/library/exceptions.rst [hunk 5]
```diff
@@ -126,13 +126,20 @@ The following exceptions are used mostly as base classes for other exceptions.
...
              raise OtherException(...).with_traceback(tb)
 
-   .. attribute:: __note__
+   .. method:: add_note(note)
 
-      A mutable field which is :const:`None` by default and can be set to a string.
-      If it is not :const:`None`, it is included in the traceback. This field can
-      be used to enrich exceptions after they have been caught.
+      Add the string ``note`` to the exception's notes which appear in the standard
+      traceback after the exception string. A :exc:`TypeError` is raise if ``note``
```
Comment (encukou) [inline Doc/library/exceptions.rst]:
```suggestion
      traceback after the exception string. A :exc:`TypeError` is raised if ``note``
```

File: Lib/test/test_exception_group.py [hunk 6]
```diff
@@ -663,6 +665,35 @@ def exc(ex):
...
+        rest.add_note("rest")
+        self.assertEqual(eg.__notes__, orig_notes + ["eg"])
+        self.assertEqual(match.__notes__, orig_notes + ["match"])
+        self.assertEqual(rest.__notes__, orig_notes + ["rest"])
+
+    def test_split_does_not_copy_non_sequence_notes(self):
+        # __notes__ should be a sequence, which is shallow copied.
+        # If it is not a sequence, the split parts don't get any notes.
+        eg = ExceptionGroup("eg", [ValueError(1), TypeError(2)])
+        eg__notes__ = 123
```
Comment (Zac-HD) [inline Lib/test/test_exception_group.py]:
```suggestion
        eg.__notes__ = 123
```

File: Objects/exceptions.c [hunk 7]
```diff
@@ -186,12 +186,61 @@ PyDoc_STRVAR(with_traceback_doc,
...
+        return NULL;
+    }
+    Py_DECREF(notes);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(add_note_doc,
+"Exception.add_note(note, replace=False) --\n\
+    add note to the exception. If replace is true, clear previous notes.");
```
Comment (iritkatriel) [inline Objects/exceptions.c]:
```suggestion
"Exception.add_note(note) --\n\
    add a note to the exception");
```

**Comment by iritkatriel** [comment]
Sorry about the noise (I messed it up multitasking). Getting there now.

**Comment by iritkatriel** [comment]
The buildbots are happy with this. Let me know if you would like to re-review, otherwise I'll merge it tomorrow or so.

File: Objects/exceptions.c [hunk 8]
```diff
@@ -186,12 +186,62 @@ PyDoc_STRVAR(with_traceback_doc,
...
+BaseException_add_note(PyObject *self, PyObject *note)
+{
+    if (!PyUnicode_Check(note)) {
+        PyErr_Format(PyExc_TypeError,
+                     "note must be a str, not '%s'",
+                     Py_TYPE(note)->tp_name);
+        return NULL;
+    }
+
+    if (!PyObject_HasAttr(self, &_Py_ID(__notes__))) {
```
Comment (serhiy-storchaka) [inline Objects/exceptions.c:206]: Do not use PyObject_HasAttr. It is broken by design.

**Review by serhiy-storchaka — COMMENTED** [review]
Opened #106033.

**Pull request merged**